contrib/cf-remote/cf_remote/packages.py

```diff
@@ -0,0 +1,188 @@
+    return numeric
+
+
+def get_default_version_data():
+    return get_latest_stable_data()
+
+
+class Artifact:
+    def __init__(self, data, filename=None):
+        if filename and not data:
+            data = {}
+            data["URL"] = "./" + filename
+            data["Title"] = ""
+            data["Arch"] = None
+            if "sparc" in filename.lower():
+                data["Arch"] = "spar"
+        self.data = data
+        self.url = data["URL"]
+        self.title = data["Title"]
+        self.arch = canonify(data["Arch"])
+
+        self.file_name = file_name(self.url)
+        self.file_extension = file_extension(self.file_name)
+
+        self.tags = ["any"]
+
+        self.add_tag(self.arch)
+        self.add_tag(self.file_extension)
+
+        look_for_tags = [
+            "Windows", "CentOS", "Red Hat", "Debian", "Ubuntu", "SLES", "Solaris", "AIX", "HPUX",
+            "HP-UX", "nova", "community", "masterfiles", "OSX", "OS X", "OS-X"
+        ]
+        for tag in look_for_tags:
+            tag = tag.lower()
+            filename = self.file_name
+            if tag in filename:
+                self.add_tag(tag)
+        if "hp-ux" in self.tags:
+            self.add_tag("hpux")
+            self.tags.remove("hp-ux")
+        if "os-x" in self.tags:
+            self.add_tag("osx")
+            self.tags.remove("os-x")
+        if "os_x" in self.tags:
+            self.add_tag("osx")
+            self.tags.remove("os_x")
+
+        parts = re.split(r"[-_.]", self.file_name)
+        #parts = self.file_name.split("-_.")
+        for part in parts:
+            part = part.strip()
+            if part == "x86":
+                continue
+            try:
+                _ = int(part)
+            except ValueError:
+                self.add_tag(part)
+
+    def add_tag(self, string):
+        assert type(string) is str
+        canonified = canonify(string)
+        if canonified not in self.tags:
+            self.tags.append(canonified)
+
+    def __str__(self):
+        return self.file_name + " ({})".format(" ".join(self.tags))
+
+
+class Release:
+    def __init__(self, data):
+        self.data = data
+        self.version = data["version"]
+        self.url = data["URL"]
+        self.lts = data["lts_branch"] if "lts_branch" in data else None
+        self.extended_data = get_json(self.url)
+        artifacts = self.extended_data["artifacts"]
+        self.artifacts = []
+        for header in artifacts:
+            for blob in artifacts[header]:
+                artifact = Artifact(blob)
+                artifact.add_tag(header)
+                self.artifacts.append(artifact)
+        self.default = False
+
+    def find(self, tags):
+        artifacts = self.artifacts
+        while len(tags) > 0:
+            tag = tags[0]
+            tag = canonify(tag)
+            remaining = filter(lambda a: tag in a.tags, artifacts)
+            remaining = list(remaining)
+            del tags[0]
```

Comment:
Why is this deleting the tags and not just iterating over them?

Reply:
Fixed.